module2/dao/src/main/java/com/epam/esm/dao/impl/GiftCertificateDaoImpl.java

```diff
@@ -0,0 +1,99 @@
+package com.epam.esm.dao.impl;
+
+import com.epam.esm.dao.GiftCertificateDao;
+import com.epam.esm.dao.entity.GiftCertificate;
+import com.epam.esm.dao.entity.Tag;
+import lombok.RequiredArgsConstructor;
+import org.springframework.jdbc.core.BeanPropertyRowMapper;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.jdbc.core.namedparam.MapSqlParameterSource;
+import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
+import org.springframework.jdbc.core.namedparam.SqlParameterSource;
+import org.springframework.jdbc.support.GeneratedKeyHolder;
+import org.springframework.jdbc.support.KeyHolder;
+import org.springframework.stereotype.Component;
+
+import java.sql.Timestamp;
+import java.time.Instant;
+import java.util.List;
+import java.util.Optional;
+
+import static com.epam.esm.dao.util.ColumnLabel.COLUMN_LABEL_CREATE_DATE;
+import static com.epam.esm.dao.util.ColumnLabel.COLUMN_LABEL_DESCRIPTION;
+import static com.epam.esm.dao.util.ColumnLabel.COLUMN_LABEL_DURATION;
+import static com.epam.esm.dao.util.ColumnLabel.COLUMN_LABEL_ID;
+import static com.epam.esm.dao.util.ColumnLabel.COLUMN_LABEL_LAST_UPDATE_DATE;
+import static com.epam.esm.dao.util.ColumnLabel.COLUMN_LABEL_NAME;
+import static com.epam.esm.dao.util.ColumnLabel.COLUMN_LABEL_PRICE;
+import static com.epam.esm.dao.util.SqlQuery.GIFT_CERTIFICATE_CREATE;
+import static com.epam.esm.dao.util.SqlQuery.GIFT_CERTIFICATE_DELETE;
+import static com.epam.esm.dao.util.SqlQuery.GIFT_CERTIFICATE_FIND_ALL;
+import static com.epam.esm.dao.util.SqlQuery.GIFT_CERTIFICATE_FIND_BY_ID;
+import static com.epam.esm.dao.util.SqlQuery.GIFT_CERTIFICATE_TAG_CREATE;
+import static com.epam.esm.dao.util.SqlQuery.GIFT_CERTIFICATE_UPDATE;
+
+@Component
+@RequiredArgsConstructor
+public class GiftCertificateDaoImpl implements GiftCertificateDao {
+
+    private final JdbcTemplate jdbcTemplate;
+    private final NamedParameterJdbcTemplate namedParameterJdbcTemplate;
+
+    @Override
+    public Optional<GiftCertificate> create(GiftCertificate giftCertificate) {
+        KeyHolder keyHolder = new GeneratedKeyHolder();
+        SqlParameterSource namedParameters = getNamedParameters(giftCertificate);
+        namedParameterJdbcTemplate.update(GIFT_CERTIFICATE_CREATE, namedParameters, keyHolder, new String[]{COLUMN_LABEL_ID});
+        Optional<GiftCertificate> createdGiftCertificate = findById(keyHolder.getKey().longValue());
+        createGiftCertificateTag(createdGiftCertificate.get().getId(), giftCertificate.getTags());
+        return createdGiftCertificate;
+    }
+
+    @Override
+    public void delete(Long id) {
+        jdbcTemplate.update(GIFT_CERTIFICATE_DELETE, id);
+    }
+
+    @Override
+    public Optional<GiftCertificate> findById(Long id) {
+        return Optional.ofNullable(jdbcTemplate.queryForObject(GIFT_CERTIFICATE_FIND_BY_ID,
+                BeanPropertyRowMapper.newInstance(GiftCertificate.class), id));
```

Comment:
from documentation:
 * Please note that this class is designed to provide convenience rather than high performance.
 * For best performance, consider using a custom {@link RowMapper} implementation.

Reply:
After discussion with mentor it was decided to use the BeanPropertyRowMapper.